tensorpipe/test/channel/channel_test.h

```diff
@@ -20,69 +20,156 @@
+    }
+
+    return v;
   }
 
-  void unwrap(void* ptr) override {
-    ASSERT_EQ(data_, ptr);
+  ~DataWrapper() {
+    if (length_ > 0) {
+      EXPECT_EQ(cudaSuccess, cudaFree(cudaPtr_));
+    }
   }
 
  private:
-  void* data_;
-  size_t size_;
+  void* cudaPtr_{nullptr};
+  size_t length_{0};
+  cudaStream_t stream_{cudaStreamDefault};
 };
 
+#endif // TENSORPIPE_HAS_CUDA
+
+template <typename TTensor>
 class ChannelTestHelper {
  public:
   virtual ~ChannelTestHelper() = default;
 
-  virtual std::shared_ptr<tensorpipe::channel::CpuContext> makeContext(
+  virtual std::shared_ptr<tensorpipe::channel::Context<TTensor>> makeContext(
       std::string id) = 0;
 
   virtual std::shared_ptr<PeerGroup> makePeerGroup() {
     return std::make_shared<ThreadPeerGroup>();
   }
-
-  virtual std::shared_ptr<DataWrapper> makeBuffer(size_t len) {
-    return std::make_shared<IdWrapper>(len);
-  }
 };
 
-class ChannelTest : public ::testing::TestWithParam<ChannelTestHelper*> {
- protected:
-  ChannelTestHelper* helper_;
-  std::shared_ptr<PeerGroup> peers_;
-
+template <typename TTensor>
+[[nodiscard]] std::pair<
+    std::future<
+        std::tuple<tensorpipe::Error, tensorpipe::channel::TDescriptor>>,
+    std::future<tensorpipe::Error>>
+sendWithFuture(
+    std::shared_ptr<tensorpipe::channel::Channel<TTensor>> channel,
+    TTensor tensor) {
+  auto descriptorPromise = std::make_shared<
+      std::promise<std::tuple<tensorpipe::Error, std::string>>>();
+  auto promise = std::make_shared<std::promise<tensorpipe::Error>>();
+  auto descriptorFuture = descriptorPromise->get_future();
+  auto future = promise->get_future();
+
+  channel->send(
+      std::move(tensor),
+      [descriptorPromise{std::move(descriptorPromise)}](
+          const tensorpipe::Error& error, std::string descriptor) {
+        descriptorPromise->set_value(
+            std::make_tuple(error, std::move(descriptor)));
+      },
+      [promise{std::move(promise)}](const tensorpipe::Error& error) {
+        promise->set_value(error);
+      });
+  return {std::move(descriptorFuture), std::move(future)};
+}
+
+template <typename TTensor>
+[[nodiscard]] std::future<tensorpipe::Error> recvWithFuture(
+    std::shared_ptr<tensorpipe::channel::Channel<TTensor>> channel,
+    tensorpipe::channel::TDescriptor descriptor,
+    TTensor tensor) {
+  auto promise = std::make_shared<std::promise<tensorpipe::Error>>();
+  auto future = promise->get_future();
+
+  channel->recv(
+      std::move(descriptor),
+      std::move(tensor),
+      [promise{std::move(promise)}](const tensorpipe::Error& error) {
+        promise->set_value(error);
+      });
+  return future;
+}
+
+template <typename TTensor>
+class ChannelTest {
  public:
-  ChannelTest() : helper_(GetParam()), peers_(helper_->makePeerGroup()) {}
+  virtual void run(ChannelTestHelper<TTensor>* helper) {
+    helper_ = helper;
+    peers_ = helper_->makePeerGroup();
```

Comment:
Doing this in the run method means that any test case that does _not_ want to use the server/client method patter must re-do this, which doesn't scale well.

Also, the fact that the default run method "forces" the server/client pattern may be odd, as users then have to override it and make it do an entirely different thing.

What we could do instead is:
- give ChannelTest an "entry point" method, which is the one called from the `TEST_P` macros below.
- give ChannelTest a pure virtual "run" method, which is called by the above entry-point method.
- have a ClientServerChannelTest subclass, which implements that run method, and in them calls the server/client methods, which are defined as purely virtual.

This way someone who wants to do a non-server/client test can subclass ChannelTest, avoid having to set up the helper_ and peers_ fields, and then implement run (without overriding an already-existing implementation).

Note that if the above suggestion (of using the constructor) works then the entry point method wouldn't be doing anything, and we could simplify it and have TEST_P directly call run.

Reply:
The problem with using a constructor (with a parameter) is that it won't be inherited. The current trade off I came up with is as follows: the helper is passed as a parameter to `run()` in every case, and when, for a specific pattern, the helper needs to be made available from other methods (which currently only happens in the ClientServerChannelTestCase), this is done in `run()`.

Reply:
Isn't there a way to have subclasses inherit constructors? Something like adding `using SuperClass;` in their body. It's not automatic (a subclass may "forget" to do so) but at least it avoids duplicating the initialization code...

Reply:
I considered the `using SuperClass::SuperClass` route but it seemed lesss convenient than the current solution (where subclasses only define test-related methods, being `run()` in the general case, or `server()/client()` when using the `ClientServer` pattern. Note that exposing the `helper` parameter of `run()` as a class member is not needed in all cases (currently only with ClientServer), and has more to do with how `ClientServerChannelTestCase` works than general "intialization of a test case".